File: backend/app/api/v1/endpoints/notifications.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_session
from app.models import User
from app.schemas.common import PaginatedResponse
from app.schemas.notification import (
    NotificationListFilter,
    NotificationRead,
    UnreadCountResponse,
)
from app.services import notification as notification_service

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=PaginatedResponse[NotificationRead])
async def list_notifications(
    filters: NotificationListFilter = Depends(),
    session: AsyncSession = Depends(get_session),
    user: User = Depends(get_current_user),
):
    return await notification_service.list_notifications(session, user, filters)


@router.get("/unread-count", response_model=UnreadCountResponse)
async def unread_count(
    session: AsyncSession = Depends(get_session),
    user: User = Depends(get_current_user),
):
    return await notification_service.get_unread_count(session, user)


@router.post("/{notification_id}/read", status_code=status.HTTP_204_NO_CONTENT)
async def mark_read(
    notification_id: int,
    session: AsyncSession = Depends(get_session),
    user: User = Depends(get_current_user),
):
    await notification_service.mark_read(session, user, notification_id)


@router.post("/mark-all-read")
async def mark_all_read(
    session: AsyncSession = Depends(get_session),
    user: User = Depends(get_current_user),
):
    n = await notification_service.mark_all_read(session, user)
    return {"updated": n}

File: frontend/src/lib/api/notifications.ts
import { apiClient } from "./client";
import type { NotificationRead, NotificationType, PaginatedResponse } from "./types";

export interface NotificationListFilter {
  is_read?: boolean;
  type?: NotificationType;
  page?: number;
  page_size?: number;
}

function buildParams(filters: NotificationListFilter): URLSearchParams {
  const p = new URLSearchParams();
  for (const [k, v] of Object.entries(filters)) {
    if (v !== undefined && v !== null && v !== "") p.append(k, String(v));
  }
  return p;
}

export const notificationsApi = {
  list: async (filters: NotificationListFilter = {}) => {
    const params = buildParams(filters);
    const res = await apiClient.get<PaginatedResponse<NotificationRead>>(
      `/notifications?${params.toString()}`,
    );
    return res.data;
  },

  unreadCount: async () => {
    const res = await apiClient.get<{ unread: number }>("/notifications/unread-count");
    return res.data.unread;
  },

  markRead: async (id: number) => {
    await apiClient.post(`/notifications/${id}/read`);
  },

  markAllRead: async () => {
    const res = await apiClient.post<{ updated: number }>("/notifications/mark-all-read");
    return res.data.updated;
  },
};
